src/main/java/ch/njol/skript/expressions/base/ChangeExpression.java

```diff
@@ -0,0 +1,73 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.expressions.base;
+
+import org.bukkit.event.Event;
+import org.eclipse.jdt.annotation.Nullable;
+
+import com.google.common.reflect.TypeToken;
+
+import ch.njol.skript.classes.Changer.ChangeMode;
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.util.coll.CollectionUtils;
+
+/**
+ * Utility class to avoid dealing with casting the delta.
+ * 
+ * @param <T> The Expression return type.
+ * @param <C> The changer type.
+ */
+public abstract class ChangeExpression<T, C> extends SimpleExpression<T> {
+
+	/**
+	 * Execute a changing of the allowed ChangeMode on this expression.
+	 * 
+	 * @param event The event where the expression is being called in.
+	 * @param values The values to set the expression with.
+	 * @param mode The ChangeMode that is being used on this expression.
+	 */
+	public abstract void changing(Event event, @Nullable C[] values, ChangeMode mode);
+
+	/**
+	 * Override this method to define the change modes for this change expression.
+	 * The change type will be the <C> generic unless {@link #acceptChange(ChangeMode)} is overridden.
+	 * In that case, this method does not need to be overridden.
+	 * 
+	 * @return An array of change modes.
+	 */
+	@Nullable
+	public ChangeMode[] getChangeModes() {
+		return null;
+	}
+
+	@Override
+	@Nullable
+	@SuppressWarnings({ "serial", "unchecked" })
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		if (!CollectionUtils.contains(getChangeModes(), mode))
+			return null;
+		return CollectionUtils.array((Class<? extends C>) new TypeToken<C>(getClass()){}.getRawType());
+	}
+
+	@Override
+	public final void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		throw new UnsupportedOperationException("ChangeExpression can only be changed from it's 'changing' method in EffChange.");
+	}
```

Comment:
why force calling changing()? Why not just make change call changing?
Then you don't need special consideration in EffChange either.
tbh i think change should still be the public api, changing should probably be protected.

Reply:
Because you cannot overload the method parameters. The changing method has the proper type array.

Reply:
> Because you cannot overload the method parameters. The changing method has the proper type array.

No, I'm saying you can just cast in change() and then call changing() from there. There's no need to make other classes call changing().

Reply:
Sure, i'll add a casting. The whole point of the class is to utilize the `changed()` method though, but if the developer wants to use this method, it can.